### PR TITLE
feat(datasets): gql for adding examples

### LINKED ISSUE #3233: [datasets][gql] add example to dataset
We want a way to explicitly pass the input, output, metadata and the optional span_id to a dataset

### PR DESCRIPTION
**Pull request by mikeldking**
resolves #3233 

```graphql
mutation addExamples($input: AddExamplesToDatasetInput!){
  addExamplesToDataset(input: $input) {
    dataset {
      name
    }
  }
}
```

### CODE REVIEW

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 1]
```diff
@@ -154,7 +159,111 @@ async def add_spans_to_dataset(
...
+        )
+
+    @strawberry.mutation
+    async def add_examples_to_dataset(
+        self, info: Info[Context, None], input: AddExamplesToDatasetInput
+    ) -> DatasetMutationPayload:
+        dataset_id = input.dataset_id
+        # Extract the span rowids from the input examples if they exit
+        span_ids = [
+            cast(GlobalID, example.span_id)
```
Comment (mikeldking) [inline src/phoenix/server/api/mutations/dataset_mutations.py:180]: is there a better way to satisfy the type checker?

File: src/phoenix/server/api/types/AddSpansToDatasetPayload.py [hunk 2]
```diff
@@ -1,8 +0,0 @@
-import strawberry
```
Comment (mikeldking) [inline src/phoenix/server/api/types/AddSpansToDatasetPayload.py:1]: @axiomofjoy nixed for now as they all are the same - happy to restore but just keeps the schema a bit tighter. WYT

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 3]
```diff
@@ -154,7 +159,111 @@ async def add_spans_to_dataset(
...
+            for example in input.examples
+            if example.span_id is not UNSET
+        ]
```
Comment (axiomofjoy) [inline src/phoenix/server/api/mutations/dataset_mutations.py:183]:
```suggestion
        span_ids = [example.span_id for example in input.examples if example.span_id]
```

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 4]
```diff
@@ -154,7 +159,111 @@ async def add_spans_to_dataset(
...
+                metadata=dataset.metadata_,
+            )
+        )
+
+    @strawberry.mutation
+    async def add_examples_to_dataset(
+        self, info: Info[Context, None], input: AddExamplesToDatasetInput
+    ) -> DatasetMutationPayload:
+        dataset_id = input.dataset_id
+        # Extract the span rowids from the input examples if they exit
```
Comment (axiomofjoy) [inline src/phoenix/server/api/mutations/dataset_mutations.py:178]:
```suggestion
        # Extract the span rowids from the input examples if they exist
```

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 5]
```diff
@@ -154,7 +159,111 @@ async def add_spans_to_dataset(
...
+        ]
+        span_rowids = {
+            from_global_id_with_expected_type(global_id=span_id, expected_type_name=Span.__name__)
+            for span_id in set(span_ids)
+        }
+        dataset_version_description = (
+            input.dataset_version_description
+            if isinstance(input.dataset_version_description, str)
+            else None
+        )
```
Comment (axiomofjoy) [inline src/phoenix/server/api/mutations/dataset_mutations.py:192]:
```suggestion
        dataset_version_description = (
            input.dataset_version_description
            if input.dataset_version_description is not UNSET
            else None
        )
```

Suggestion: This pattern doesn't require us to know the type of the field.

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 6]
```diff
@@ -154,7 +159,111 @@ async def add_spans_to_dataset(
...
+                insert(models.DatasetVersion)
+                .values(
+                    dataset_id=dataset_rowid,
+                    description=dataset_version_description,
+                    metadata_=dataset_version_metadata or {},
+                )
+                .returning(models.DatasetVersion.id)
+            )
+            print("dataset_version_rowid" + str(dataset_version_rowid))
+            print("span_ids" + str(span_ids))
```
Comment (axiomofjoy) [inline src/phoenix/server/api/mutations/dataset_mutations.py:216]: Can we nix these print statement?

Comment (axiomofjoy) [reply src/phoenix/server/api/mutations/dataset_mutations.py:216]: Cool thing I recently learned, you can use `print(f"{dataset_version_rowid=}")` as a shorthand.

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 7]
```diff
@@ -154,7 +159,111 @@ async def add_spans_to_dataset(
...
+                await session.execute(
+                    select(models.Span.id)
+                    .select_from(models.Span)
+                    .where(models.Span.id.in_(span_rowids))
+                )
+            ).all()
+            # Just validate that the number of spans matches the number of span_ids
+            # to ensure that the span_ids are valid
+            assert len(spans) == len(span_rowids)
+            print("spans" + str(spans))
```
Comment (axiomofjoy) [inline src/phoenix/server/api/mutations/dataset_mutations.py:227]: Same.

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 8]
```diff
@@ -154,7 +159,111 @@ async def add_spans_to_dataset(
...
+                raise ValueError(
+                    f"Unknown dataset: {dataset_id}"
+                )  # todo: implement error types https://github.com/Arize-ai/phoenix/issues/3221
+            dataset_version_rowid = await session.scalar(
+                insert(models.DatasetVersion)
+                .values(
+                    dataset_id=dataset_rowid,
+                    description=dataset_version_description,
+                    metadata_=dataset_version_metadata or {},
+                )
```
Comment (axiomofjoy) [inline src/phoenix/server/api/mutations/dataset_mutations.py:212]:
```suggestion
        dataset_version_metadata = input.dataset_version_metadata or {}
        dataset_rowid = from_global_id_with_expected_type(
            global_id=dataset_id, expected_type_name=Dataset.__name__
        )
        async with info.context.db() as session:
            if (
                dataset := await session.scalar(
                    select(models.Dataset).where(models.Dataset.id == dataset_rowid)
                )
            ) is None:
                raise ValueError(
                    f"Unknown dataset: {dataset_id}"
                )  # todo: implement error types https://github.com/Arize-ai/phoenix/issues/3221
            dataset_version_rowid = await session.scalar(
                insert(models.DatasetVersion)
                .values(
                    dataset_id=dataset_rowid,
                    description=dataset_version_description,
                    metadata_=dataset_version_metadata,
                )
```

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 9]
```diff
@@ -154,7 +159,111 @@ async def add_spans_to_dataset(
...
+                .returning(models.DatasetVersion.id)
+            )
+            print("dataset_version_rowid" + str(dataset_version_rowid))
+            print("span_ids" + str(span_ids))
+            spans = (
+                await session.execute(
+                    select(models.Span.id)
+                    .select_from(models.Span)
```
Comment (axiomofjoy) [inline src/phoenix/server/api/mutations/dataset_mutations.py:220]: Non-blocking: This `select_from` is probably unnecessary.

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 10]
```diff
@@ -154,7 +159,111 @@ async def add_spans_to_dataset(
...
+                    .where(models.Span.id.in_(span_rowids))
+                )
+            ).all()
+            # Just validate that the number of spans matches the number of span_ids
+            # to ensure that the span_ids are valid
+            assert len(spans) == len(span_rowids)
+            print("spans" + str(spans))
+            # Create a map of span rowids to
```
Comment (axiomofjoy) [inline src/phoenix/server/api/mutations/dataset_mutations.py:228]: Incomplete comment. Looks out of date.

File: src/phoenix/server/api/mutations/dataset_mutations.py [hunk 11]
```diff
@@ -154,7 +159,111 @@ async def add_spans_to_dataset(
...
+                                global_id=cast(GlobalID, example.span_id),
+                                expected_type_name=Span.__name__,
+                            )
+                            if example.span_id is not UNSET
+                            else None,
+                        }
+                        for example in input.examples
+                    ],
+                )
+            ).all()
```
Comment (axiomofjoy) [inline src/phoenix/server/api/mutations/dataset_mutations.py:246]:
```suggestion
            dataset_example_rowids = (
                await session.scalars(
                    insert(DatasetExample).returning(DatasetExample.id),
                    [
                        {
                            DatasetExample.dataset_id.key: dataset_rowid,
                            DatasetExample.span_rowid.key: from_global_id_with_expected_type(
                                global_id=example.span_id,
                                expected_type_name=Span.__name__,
                            )
                            if example.span_id
                            else None,
                        }
                        for example in input.examples
                    ],
                )
            ).all()
```

File: src/phoenix/server/api/input_types/DatasetExampleInput.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+from strawberry import UNSET
+from strawberry.relay import GlobalID
+from strawberry.scalars import JSON
+
+
+@strawberry.input
+class DatasetExampleInput:
+    input: JSON
+    output: JSON
+    metadata: JSON
```
Comment (axiomofjoy) [inline src/phoenix/server/api/input_types/DatasetExampleInput.py:13]: Should we allow this to be `UNSET`?

File: src/phoenix/server/api/types/AddSpansToDatasetPayload.py (hunk 2 above)
Comment (axiomofjoy) [reply src/phoenix/server/api/types/AddSpansToDatasetPayload.py:1]: Not sure I have a strong opinion here. What does GitHub do?

**Review by axiomofjoy — APPROVED** [review]

File: src/phoenix/server/api/types/AddSpansToDatasetPayload.py (hunk 2 above)
Comment (mikeldking) [reply src/phoenix/server/api/types/AddSpansToDatasetPayload.py:1]: I think for now let's keep it shorter.

File: src/phoenix/server/api/mutations/dataset_mutations.py (hunk 6 above)
Comment (mikeldking) [reply src/phoenix/server/api/mutations/dataset_mutations.py:211]: Oh neat

File: src/phoenix/server/api/mutations/dataset_mutations.py (hunk 8 above)
Comment (mikeldking) [reply src/phoenix/server/api/mutations/dataset_mutations.py:207]: I made this required for now

File: src/phoenix/server/api/input_types/DatasetExampleInput.py (hunk 12 above)
Comment (mikeldking) [reply src/phoenix/server/api/input_types/DatasetExampleInput.py:13]: I'm leaving it for now as required

File: src/phoenix/server/api/types/AddSpansToDatasetPayload.py (hunk 2 above)
Comment (axiomofjoy) [reply src/phoenix/server/api/types/AddSpansToDatasetPayload.py:1]: Sounds good. For reference, looks like GitHub creates output types for each mutation.

Screenshot 2024-05-22 at 10 45 30 AM

Screenshot 2024-05-22 at 10 45 42 AM

**Pull request merged**